third_party/move/move-vm/types/src/loaded_data/runtime_access_specifier.rs

```diff
@@ -0,0 +1,536 @@
+//! function's specifier. The join operator is complete (no approximation). A further
+//! operator `AccessSpecifier::subsumes` allows to test whether one specifier
+//! allows all the access of the other. This used to abort execution if a function
+//! is entered which declares accesses not allowed by the context. However, the
+//!`subsumes` function is incomplete. This is semantically sound since
+//! if subsume is undecided, abortion only happens later at the time of actual access
+//! instead of when the function is entered.
+//!
+//! The `join` operation attempts to simplify the resulting access specifier, making
+//! access checks faster and keeping memory use low. This is only implemented for
+//! inclusions, which are fully simplified. Exclusions are accumulated.
+//! There is potential for optimization by simplifying exclusions but since those are effectively
+//! negations, such a simplification is not trivial and may require recursive specifiers, which
+//! we like to avoid.
+
+use crate::{
+    loaded_data::runtime_types::{StructIdentifier, Type},
+    values::{Reference, SignerRef, Value},
+};
+use itertools::Itertools;
+use move_binary_format::{
+    errors::{PartialVMError, PartialVMResult},
+    file_format::{AccessKind, LocalIndex},
+};
+use move_core_types::{
+    account_address::AccountAddress, language_storage::ModuleId, vm_status::StatusCode,
+};
+use std::{fmt, fmt::Debug};
+
+/// Represents an access specifier.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub enum AccessSpecifier {
+    /// Universal access granted
+    Any,
+    /// A constraint in normalized form: `Constraint(inclusions, exclusions)`.
+    /// The inclusions are a _disjunction_ and the exclusions a _conjunction_ of
+    /// access clauses. An access is valid if it is enabled by any of the
+    /// inclusions, and not enabled for each of the exclusions.
+    Constraint(Vec<AccessSpecifierClause>, Vec<AccessSpecifierClause>),
+}
+
+/// Represents an access specifier clause
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub struct AccessSpecifierClause {
+    pub kind: AccessKind,
+    pub resource: ResourceSpecifier,
+    pub address: AddressSpecifier,
+}
+
+/// Represents a resource specifier.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub enum ResourceSpecifier {
+    Any,
+    DeclaredAtAddress(AccountAddress),
+    DeclaredInModule(ModuleId),
+    Resource(StructIdentifier),
+    ResourceInstantiation(StructIdentifier, Vec<Type>),
+}
+
+/// Represents an address specifier.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub enum AddressSpecifier {
+    Any,
+    Literal(AccountAddress),
+    /// The `Eval` specifier represents a value dependent on a parameter of the
+    /// current function. Once address specifiers are instantiated in a given
+    /// caller context it is replaced by a literal.
+    Eval(AddressSpecifierFunction, LocalIndex),
```

Comment:
So here we're basically saying that an address specifier function must be a function over a single local index. Would this be a bit restrictive for potential future use cases?

Wondering if we should instead have
```
pub enum AddressSpecifier {
    Eval(AddressSpecifierFunction),
}

pub enum AddressSpecifierFunction {
    SignerAddress(LocalIndex),
    ObjectAddress(LocalIndex),
}
```

Reply:
This is an interesting generalization, however, the decision its one local (in fact a parameter) is already made in the type checker, so this change would require more work in the front end. We can make this change potentially later as long as V7 file format isn't default yet.

Reply:
> so this change would require more work in the front end

Hmm.. are we sharing the AddressSpecifier with the frontend? If not then it might be trivial to generalize it.

Agreed that there is no rush to generalize it though. Runtime representations like this should not have backward compatibility problems in general.